File: tests/mozmill/tests/testLinks/testHTMLAnchorElement/testOpenInNewTab.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

var {assert, expect} = require("../../../../../../../lib/assertions");
var prefs = require("../../../../../../lib/prefs");
var tabs = require("../../../../../../lib/tabs");
var utils = require("../../../../../../../lib/utils");

var rpConst = require("../../../lib/constants");

const TEST_URL = "http://www.maindomain.test/link_1.html";


var setupModule = function(aModule) {
  aModule.controller = mozmill.getBrowserController();

  aModule.tabBrowser = new tabs.tabBrowser(aModule.controller);
  aModule.tabBrowser.closeAllTabs();

  prefs.preferences.setPref(rpConst.PREF_DEFAULT_ALLOW, false);
}

var teardownModule = function(aModule) {
  prefs.preferences.clearUserPref(rpConst.PREF_DEFAULT_ALLOW);
  utils.closeContentAreaContextMenu(aModule.controller);
  aModule.tabBrowser.closeAllTabs();
}


var testOpenInBackgroundTab = function() {
  controller.open(TEST_URL);
  controller.waitForPageLoad();

  let link = getLink();
  let linkURL = link.getNode().href;

  let i = 1;
  while (true === openNextTab(i, link)) {
    // Check that i+1 tabs are open
    assert.waitFor(function () {
      return tabBrowser.length === (i + 1);
    }, "Tab " + (i + 1) + " opened.");
    ++i;
  }

  assertCorrectLocations(linkURL);
  assertNoRedirects();
}


/**
 * @return {MozMillElement} The link to click on.
 */
var getLink = function() {
  let links = controller.window.content.document.getElementsByTagName("a");
  assert.notEqual(links.length, 0, "A link has been found on the test page.");

  return findElement.Elem(links[0]);
}

/**
 * Opens the next tab.
 * @return {boolean}
 *         true if a new tab has been opened.
 *         false if no tab needs to open anymore.
 */
var openNextTab = function(i, link) {
  switch (i) {
    case 1:
      // Open another tab by middle-clicking on the link
      tabBrowser.openTab({method: "middleClick", target: link});
      return true;
      break;

    case 2:
      // Open link via context menu in a new tab:
      tabBrowser.openTab({method: "contextMenu", target: link});
      return true;
      break;

    default:
      return false;
      break;
  }
}

var assertCorrectLocations = function(linkURL) {
  for (let index = 1; index < tabBrowser.length; ++index) {
    tabBrowser.selectedIndex = index;
    assert.equal(controller.tabs.activeTab.location.href, linkURL,
        "The location in the new tab is correct.");
  }
}


/**
 * Assert that the link clicks have not been detected as redirects.
 */
var assertNoRedirects = function() {
  for (let index = 0; index < tabBrowser.length; ++index) {
    var panel = tabBrowser.getTabPanelElement(index,
        '/{"value":"' + rpConst.REDIRECT_NOTIFICATION_VALUE + '"}');
    assert.ok(false === panel.exists(),
        "Following the link didn't cause a redirect");
  }
}

File: tests/marionette/tests/links/html_anchor_element/test_open_in_new_tab.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

from rp_ui_harness import RequestPolicyTestCase
from marionette_driver.marionette import Actions


TEST_URL = "http://www.maindomain.test/link_1.html";
PREF_DEFAULT_ALLOW = "extensions.requestpolicy.defaultPolicy.allow"


class TestOpenInNewTab(RequestPolicyTestCase):
    def setUp(self):
        RequestPolicyTestCase.setUp(self)
        self.prefs.set_pref(PREF_DEFAULT_ALLOW, False);

    def test_open_in_new_tab(self):
        with self.marionette.using_context("content"):
            # load the test url
            self.marionette.navigate(TEST_URL)
            # get the link and its url
            link = self.marionette.find_element("tag name", "a")
            link_url = link.get_attribute("href")

        with self.marionette.using_context("chrome"):
            # Open the link in a new tab, using all possible methods
            # sequentially.
            for _ in self.open_link_multi(link):
                self.assertEqual(len(self.browser.tabbar.tabs), 2,
                                 "A new tab has been opened.")

                [origin_tab, dest_tab] = self.browser.tabbar.tabs
                self.tabs.wait_until_loaded(dest_tab)

                # checks in the origin's tab
                origin_tab.switch_to()
                self.assertFalse(self.redir.is_shown(),
                                 ("Following the link didn't cause a "
                                  "redirect in the origin tab."))

                # checks in the destination's tab
                dest_tab.switch_to()
                self.assertEqual(dest_tab.location, link_url,
                                 "The location in the new tab is correct.")
                self.assertFalse(self.redir.is_shown(),
                                 ("Following the link didn't cause a "
                                  "redirect in the destination tab."))

                dest_tab.close()

    def open_link_multi(self, link):
        """Open a link in new tabs using different methods."""

        # METHOD #1: middle click
        with self.marionette.using_context("content"):
            Actions(self.marionette).click(link, 1).perform()
        yield

        # METHOD #2: context menu
        self.ctx_menu.select_entry("context-openlinkintab", link)
        # TODO: Use the "tabs" library as soon as it has been ported
        #       to Marionette, see Mozilla Bug 1121725.
        #       The mozmill code to open the link in a new tab was:
        #       ```
        #       tabBrowser.openTab({method: "contextMenu", target: link});
        #       ```
        yield
